### PR TITLE
kubeadm: refactor config

### PR DESCRIPTION
**Pull request by mikedanese**
1) break object into substructures
2) seperate a config object for master and node
3) centralize defaulting and validation

Hacked til it compiled. Have not done 3 yet.

Step one of #33715

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33728)

### CODE REVIEW

File: cmd/kubeadm/app/api/env.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+// TODO(phase2) use componentconfig
+// we need some params for testing etc, let's keep these hidden for now
+func GetEnvParams() map[string]string {
+
+	envParams := map[string]string{
+		// TODO(phase1+): Mode prefix and host_pki_path to another place as constants, and use them everywhere
+		// Right now they're used here and there, but not consequently
+		"kubernetes_dir":     "/etc/kubernetes",
```
Comment (mikedanese) [inline cmd/kubeadm/app/api/env.go]: Move into config type and default in centralized defaulting but still pull from env until file based config is available.

File: cmd/kubeadm/app/api/types.go [hunk 2]
```diff
@@ -16,89 +16,54 @@ limitations under the License.
...
+	ExternalDNSNames   []string
 }
 
-const (
-	DefaultServiceDNSDomain   = "cluster.local"
-	DefaultServicesCIDRString = "100.64.0.0/12" // Carrier-grade NAT range (RFC 6598)
-	DefaultKubernetesVersion  = "v1.4.0"
-)
+type Networking struct {
+	ServiceSubnet string
```
Comment (luxas) [inline cmd/kubeadm/app/api/types.go]: Why not use `net.IPNet` here for automatic validation+

File: cmd/kubeadm/app/cmd/init.go [hunk 3]
```diff
@@ -63,89 +62,58 @@ func NewCmdInit(out io.Writer, s *kubeadmapi.KubeadmConfig) *cobra.Command {
...
 	cmd.PersistentFlags().StringSliceVar(
-		&s.InitFlags.API.ExternalDNSNames, "api-external-dns-names", []string{},
+		&s.API.ExternalDNSNames, "api-external-dns-names", []string{},
 		"The DNS names to advertise, in case you have configured them yourself",
 	)
-	cmd.PersistentFlags().IPNetVar(
-		&s.InitFlags.Services.CIDR, "service-cidr", *kubeadmapi.DefaultServicesCIDR,
+	cmd.PersistentFlags().StringVar(
+		&s.Networking.ServiceSubnet, "service-cidr", "", //*kubeadmapi.DefaultServicesCIDR,
 		`Use alterantive range of IP address for service VIPs, defaults to `+
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: alternative

File: cmd/kubeadm/app/api/types.go (hunk 2 above)
Comment (mikedanese) [inline cmd/kubeadm/app/api/types.go]: I'm getting this ready to become an API group which won't support that type. I suppose I can transistion it in a follow up.

Comment (luxas) [inline cmd/kubeadm/app/api/types.go]: Aha, so api groups don't support `IPNet`?
Thanks, didn't know that

File: cmd/kubeadm/app/api/env.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"fmt"
+	"os"
+	"runtime"
+	"strings"
+)
+
+// TODO(phase2) use componentconfig
+// we need some params for testing etc, let's keep these hidden for now
+func GetEnvParams() map[string]string {
```
Comment (krousey) [inline cmd/kubeadm/app/api/env.go]: Can we kill `map[string]string` and just use a struct. That way the string names only exist in one place and have only one place to be mis-typed.

Comment (luxas) [inline cmd/kubeadm/app/api/env.go]: I also think we should get rid of this now while refactoring

**Comment by errordeveloper** [comment]
Mike, I am planing to take a look at this, haven't had the time yet...

**Comment by errordeveloper** [comment]
Yes, none of environment variables are documented, please feel free to do
anything, as long some roughly similar options are available :)

File: cmd/kubeadm/app/api/env.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package api
+
+import (
+	"fmt"
+	"os"
+	"runtime"
+	"strings"
+)
+
+// TODO(phase2) use componentconfig
```
Comment (errordeveloper) [inline cmd/kubeadm/app/api/env.go]: please remove the TODO 👍

File: cmd/kubeadm/app/api/env.go (hunk 1 above)
Comment (errordeveloper) [inline cmd/kubeadm/app/api/env.go]: Are you planning to implement it here, or there is some other dependency?

**Review by errordeveloper — COMMENTED** [review]
Happy to merge as soon as it's functional, and we can   extend it in another step, if you like.

**Comment by luxas** [comment]
Please gofmt and fix boilerplate for cmd/kubeadm/app/api/env.go if you think you're ready

**Comment by mikedanese** [comment]
This is in decent shape and getting it in in it's current form will allow us to continue working on this and other stuff without rebase hell. @krousey has a PR that will probably conflict with this heavily.

File: cmd/kubeadm/app/api/env.go (hunk 5 above)
Comment (mikedanese) [inline cmd/kubeadm/app/api/env.go]: I don't think this is compelte until we have a real apigroup. I'll remove once that's done.

**Comment by mikedanese** [comment]
Looks like a non blockinf flake. @errordeveloper @luxas PTAL

**Comment by luxas** [comment]
I'm not able to review this in detail right now, but a quick look at it says "looks good to me so we can get it merged in order to not have a rebase hell later".

The next step is probably integrating the `envParams` map into the API, and then convert this API into an alpha componentconfig API.

**Pull request merged**